### PR TITLE
Implement proper in-memory ephemeral localStorage support.

### PR DESCRIPTION
**Pull request by goodov**
Implemented localStorage via LocalStorageImpl using in-memory mode. Everything should work as before, but the underlying logic now allows us to share ```localStorage``` between frames as it's originally intended. A test was added to ensure this now works properly (it fails on the previous implementation).

Resolves https://github.com/brave/brave-browser/issues/16920

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: browser/ephemeral_storage/sources.gni [hunk 1]
```diff
@@ -0,0 +1,14 @@
+brave_browser_ephemeral_storage_sources = [
+  "//brave/browser/ephemeral_storage/ephemeral_storage_tab_helper.cc",
+  "//brave/browser/ephemeral_storage/ephemeral_storage_tab_helper.h",
+]
+
+brave_browser_ephemeral_storage_deps = [
+  "//base",
+  "//chrome/browser/profiles",
```
Comment (iefremov) [inline browser/ephemeral_storage/sources.gni:8]: AFAIK it doesn't add a circular dep, so these sources/deps can stay in build.gn

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.cc [hunk 2]
```diff
@@ -47,6 +50,23 @@ std::string StringToSessionStorageId(const std::string& string,
   return hash;
 }
 
+class EphemeralStorageOriginsSourceImpl
```
Comment (iefremov) [inline browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:53]: a comment?

File: components/content_settings/renderer/brave_content_settings_agent_impl.cc [hunk 3]
```diff
@@ -133,46 +133,53 @@ void BraveContentSettingsAgentImpl::DidNotAllowScript() {
   ContentSettingsAgentImpl::DidNotAllowScript();
 }
 
-bool BraveContentSettingsAgentImpl::UseEphemeralStorageSync(
+blink::WebSecurityOrigin
+BraveContentSettingsAgentImpl::GetEphemeralStorageOriginSync(
     StorageType storage_type) {
   if (!base::FeatureList::IsEnabled(net::features::kBraveEphemeralStorage))
-    return false;
+    return blink::WebSecurityOrigin();
```
Comment (iefremov) [inline components/content_settings/renderer/brave_content_settings_agent_impl.cc:140]: mb just `return {};`

File: chromium_src/content/public/browser/tld_ephemeral_lifetime.h [hunk 4]
```diff
@@ -41,14 +43,26 @@ class CONTENT_EXPORT TLDEphemeralLifetime
  public:
   using OnDestroyCallback = base::OnceCallback<void(const std::string&)>;
 
+  class EphemeralStorageOriginsSource {
+   public:
+    virtual ~EphemeralStorageOriginsSource() = default;
+
+    virtual std::vector<url::Origin> TakeEphemeralStorageOpaqueOrigins(
```
Comment (iefremov) [inline chromium_src/content/public/browser/tld_ephemeral_lifetime.h:50]: needs some explanations

**Review by iefremov — APPROVED** [review]
I *think* I understand how it works and it looks good, also I've run manual tests with these changes

File: browser/ephemeral_storage/sources.gni (hunk 1 above)
Comment (goodov) [reply browser/ephemeral_storage/sources.gni:8]: sadly it adds.

**Review by mariospr — APPROVED** [review]
LGTM for chromium_src changes

**Comment by o9o** [comment]
This has broken my extension: https://chrome.google.com/webstore/detail/dark-mode-night-reader/hmafjphdklmdjfcnljjeonfpgafanjjc

It stores per-website state in the page's local storage. Brave is now forgetting it when the tab is closed.

**Pull request merged**